polygeist/include/mlir/Dialect/Polygeist/Analysis/ReachingDefinitionAnalysis.h

```diff
@@ -15,10 +15,58 @@
 
 #include "mlir/Analysis/DataFlow/DenseAnalysis.h"
 #include "mlir/Dialect/Polygeist/Utils/AliasUtils.h"
+#include <set>
+#include <variant>
 
 namespace mlir {
 namespace polygeist {
 
+/// Represents a definition that is unknown (this is a singleton).
+class InitialDefinition {
+  friend raw_ostream &operator<<(raw_ostream &, const InitialDefinition &);
+
+public:
+  InitialDefinition(InitialDefinition &) = delete;
+  bool operator=(const InitialDefinition &) = delete;
+  static InitialDefinition *getInstance();
+
+private:
+  InitialDefinition() = default;
+  static InitialDefinition *singleton;
+};
```

Comment:
```suggestion
/// Represents a definition that is unknown (this is a singleton).
class InitialDefinition {
  friend raw_ostream &operator<<(raw_ostream &, const InitialDefinition &);

public:
  InitialDefinition(InitialDefinition &) = delete;
  InitialDefinition operator=(const InitialDefinition &) = delete;
  InitialDefinition(InitialDefinition &&) = delete;
  InitialDefinition operator=(InitialDefinition &&) = delete;
  static constexpr InitialDefinition get() { return instance; }

private:
  InitialDefinition() = default;
  static constexpr InitialDefinition singleton;
};
```

Reply:
Avoid memory allocation (leaking) using `constexpr` singleton

Reply:
Can we use a different name, e.g., `Undef`?

Reply:
Yes, sorry. You should declare `get` in the `.h` and have the singleton declared as a global `constexpr` variable in the `.cpp` file and `get` defined there. That should work.

Reply:
I wouldn't because (in the .cpp file) this definition:
```
static constexpr InitialDefinition instance;
```
would not compile given that the constructor for `InitialDefinition` is private.

Reply:
This is now gone @victor-eds, I changed the InitialDefinition class to an "empty" class.